src/core/observer/index.js

```diff
@@ -189,7 +189,7 @@ export function defineReactive (
  * already exist.
  */
 export function set (target: Array<any> | Object, key: any, val: any): any {
-  if (Array.isArray(target) && typeof key === 'number') {
+  if (Array.isArray(target) && !isNaN(parseFloat(key))) {
```

Comment:
I think something like a double `typeof...` (+ `key !== ''`) or `key !== '' && typeof Number(key) === 'number'` would be better. `parseFloat` looks like an overkill and `isNaN` is not supported on IE. wdyt @defcc ?